### ISSUE TITLE
[MNG-6084] Support JSR 250 annotations

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7658]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** opened **[MNG-6084](https://issues.apache.org/jira/browse/MNG-6084?redirect=false)** and commented

JSR330 ( `@Named`, `@Singeton`, etc) currently supported for plugin development.

Would like to see JSR250 as well(`@PostConstruct`, etc)

Detail discussion is at  http://maven.40175.n5.nabble.com/Maven-Plugin-and-JSR330-td5879508.html

---

**Affects:** 3.3.9

**Issue Links:**
- [MNGSITE-506](https://issues.apache.org/jira/browse/MNGSITE-506) Document JSR 250 annotation
 (_**"causes"**_)
- [MNG-6298](https://issues.apache.org/jira/browse/MNG-6298) 3.5.2: ClassNotFoundException: javax.annotation.security.RolesAllowed
 (_**"breaks"**_)

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=67636510)
- [GitHub Pull Request #15
](https://github.com/apache/maven-integration-testing/pull/15)
- [GitHub Pull Request #98
](https://github.com/apache/maven/pull/98)

0 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Stuart McCulloch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mcculls)** commented

Support for JSR250 lifecycle annotations can be enabled in the `ContainerConfiguration`, configured in
`maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java`:

```
ContainerConfiguration cc = new DefaultContainerConfiguration()
    .setClassWorld( cliRequest.classWorld )
    .setRealm( containerRealm )
    .setClassPathScanning( PlexusConstants.SCANNING_INDEX )
    .setAutoWiring( true )
    .setJSR250Lifecycle( true )        // <--- enable support for @PostConstruct and @PreDestroy
    .setName( "maven" );
```

You may also want to add `javax.annotation` to the list of packages exported to plugins from Maven core, defined in
`maven-core/src/main/resources/META-INF/maven/extension.xml`:

```
<!-- javax.annotation (JSR-250) -->
<exportedPackage>javax.annotation.*</exportedPackage>
```

otherwise it will use the package from the current JRE, which might not have it (ie. Java 7 and earlier) or might hide it by default (Jigsaw)

**Comment by jira-importer** [comment]
**[Stuart McCulloch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mcculls)** commented

Note: any mojos which have `@PreDestroy` methods will be disposed as soon as they're done (Maven tells the container about this explicitly in DefaultMavenPluginManager.java line 759) whereas leftover components which have `@PreDestroy` methods will be disposed at the end of the build. (The next container release should be more aggressive about disposing of non-mojo components.)

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

`@Stuart`,  could you elaborate on 'next container release'? where is it?  Can we file a JIRA as a reminder?

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

Commit suggested changes at https://github.com/dantran/maven.git

with a test case  at https://github.com/dantran/maven-integration-testing.git

so far my local test does not see PostConstruct working yet

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

JSR 250 integration tests are now working.  the commits are squashed at both maven and maven integration repo

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

Special thanks to Stuart for guiding me thru this enhancement

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

Maven 3.4.0 has been dropped. See [this thread](https://mail-archives.apache.org/mod_mbox/maven-dev/201701.mbox/%3CCA%2BnPnMx-e7kGYy3Hp87v8hLGdhp1q%3DtKLx_6QuZ4kGUqHEBGcw%40mail.gmail.com%3E) for more details.

This issue will need to be re-scheduled for a Maven release in the (hopefully near) future.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [a7dddcb87626fec256f1b8ccbe957f94cdf401e4](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commit;h=a7dddcb87626fec256f1b8ccbe957f94cdf401e4).
